include/clang/AST/Expr.h

```diff
@@ -4514,6 +4514,120 @@ class GenericSelectionExpr : public Expr {
   friend class ASTStmtReader;
 };
 
+/// \brief Represents a Checked C bounds expression.
+class BoundsExpr : public Expr {
```

Comment:
Should this derive from Expr?  If it's not a C++ evaluatable expression (which yields a value at compile time or runtime), and if only something that can be attached to a variable or field, then could it be a standalone class with no parents?  I see that Expr derives from Stmt; perhaps it could be a peer of Stmt, then, as a toplevel class in the AST class hierarchy.

I ask not just to make sure that we're appropriately representing it, but also to avoid cases where someone accidentally could stick a bounds expression somewhere they ought not to (e.g. the condition expression of an if statement).  Even if the parser wouldn't do that, that wouldn't stop someone from programmatically inserting one there in an AST.